docs/dev/core/internal/upload.md

```diff
@@ -7,6 +7,62 @@ The Glean SDK leverages the available platform capabilities to implement any net
 Glean core controls all upload and coordinates the platform side with its own internals.
 All language bindings implement ping uploading around a common API and protocol.
 
+## The upload module in the language bindings
+This paragraph illustrates the recommended structure of the upload module implementation in the language bindings (regardless of the target language), in order to allow greater flexibility with respect to potential future requirements and features.
+
+```mermaid
+classDiagram
+    class UploadResult {
+        ~ToFFI()* int
+    }
+
+    class HttpResponse {
+        int statusCode
+        ~ToFFI() int
+    }
+
+    class UnrecoverableFailure {
+        ~ToFFI() int
+    }
+
+    class RecoverableFailure {
+        ~ToFFI() int
+    }
+
+    class PingUploader {
+        <<interface>>
+        +Upload() UploadResult
+    }
+
+    class BaseUploader {
+        +BaseUploader(PingUploader)
+        ~TriggerUploads()
+        ~CancelUploads()
+    }
+
+    class HttpUploader {
+        +Upload() UploadResult
+    }
+
+    UploadResult <|-- HttpResponse
+    UploadResult <|-- UnrecoverableFailure
+    UploadResult <|-- RecoverableFailure
+    PingUploader <|-- HttpUploader
+    PingUploader o-- BaseUploader
+
+```
+
+* The `PingUploader` interface describes the contract between the `BaseUploader` and the SDK or user-provided upload modules.
+* There is a `BaseUploader` component that is responsible for interfacing with the lower level FFI `get_upload_*` calls and dealing with the logic in a platform-coherent way.
+  * Every Glean instance will always have a single `BaseUploader` instance.
+  * The `BaseUploader` is fed, at Glean initialization, with an instance of an implementation of the `PingUploader` interface.
+  * Whenever `BaseUploader` thinks it should perform an upload, it will call the provided instance of the `PingUploader` interface and call `upload` with the data it's getting from the glean-core/FFI.
+  * Any throttling happens at this layer!
```

Comment:
```suggestion
  * Any throttling happens at this layer.
```
Any specific reason for the exclamation here? One more thing, do you think it is necessary to point out that the core is what should orchestrate the throttling and the binding will just abide to what the core is telling it to do?

Reply:
Oh, I tend to get overexcited with punctuation! Sorry about that!

Yes, it makes sense to clarify the throttling part. Let me add that.